crypto.cpp

```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
     } \
 }
 
+static
+VOID
+OvpnCryptoMakeEpochNonce(UCHAR *epochIv, UINT64 packet_id_net, UCHAR *nonce)
+{
+    // first 8 bytes of IV (aka nonce) is pktid_net XOR 8 bytes of epoch IV
+    UINT64 iv0;
+    RtlCopyMemory(&iv0, epochIv, sizeof(iv0));
+    iv0 ^= packet_id_net;
+    RtlCopyMemory(nonce, &iv0, sizeof(iv0));
+
+    // last 4 bytes of IV are from epoch IV
```

Comment:
Do you mean the epoch itself?

Reply:
Epoch's implicit IV. As RFC says, `implicit IV belonging to the epoch`. Clarified in comments.